### PR TITLE
Externalize kvm agent storage reboot configuration

### PR DESCRIPTION
**Pull request by GutoVeronezi**
### Description
On KVMHAMonitor, when there is an inconsistency on the heartbeat's file or heartbeat timeout is extrapolated several times, the host is restarted. The host restarting amid several operations might bring inconsistency problems; therefore, it is interesting to let operators choose if they want the host restart or not on the hearbeat's timeout or storage errors. 

This PR intends to externalize the heartbeat reboot configuration on KVMHAMonitor. The default value is `true`; therefore, the current behavior is maintained.

This PR depends on [PR#4585](https://github.com/apache/cloudstack/pull/4585).

### Types of changes
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functioanlity)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale
- [x] Major
- [ ] Minor

### How Has This Been Tested?
It has been tested locally on a test lab.
1. On agent, I changed my log4j to debug mode.
2. I added `hearbeat.update.timeout=10` setting to `agent.properties` to force a timeout on hearbeat.
3. I added `reboot.host.and.alert.management.on.hearbeat.timeout` setting to `agent.properties` and changed it some times;
4. I restarted agent and watched the log.

### CODE REVIEW

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java
Comment (RodrigoDLopez) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:38]: And if we start to normalize these variable names in this file?
Removing these underscore at the start of any variable. at least where some changes were needed

**Review by RodrigoDLopez — CHANGES_REQUESTED** [review]
Hey @GutoVeronezi  nice feature ...

I was thinking and if we change the name of these files, put something more accurate or self explainable?

1. AgentProperty.java => AgentProperties.java
since we will work to bring all available properties to this file, right?

2. AgentPropertyFile.java => AgentPropertiesFileHandler.java
this class sounds like a handler or facility to me, am I right?

Bearing in mind that we are looking to standardize and normalize things here. What do you think of normalizing the KVMHAMonitor.java file, at least on the lines where some changes were needed.
> String _hostIP => String hostIP

My eyes bleed when I see things like that in the code.

**Comment by GutoVeronezi** [comment]
@RodrigoDLopez sure, the new files names will be more intuitive.
I'll do it.

About the normalizations of the names, I understand what you mean, I'll do it too.

**Comment by DaanHoogland** [comment]
good work @GutoVeronezi looking forward to see this merged, but I am going to put my foot down for very extensive testing. :+1:

**Review by RodrigoDLopez — APPROVED** [review]

**Comment by RodrigoDLopez** [comment]
nice job @GutoVeronezi 
Code LGTM.

**Comment by GutoVeronezi** [comment]
Can anyone review this PR?

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java
Comment (DaanHoogland) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:38]: you mean
```suggestion
    private final Map<String, NfsStoragePool> storagePool = new ConcurrentHashMap<>();
```
@RodrigoDLopez ;)
(will need work in the rest of the file!!!)

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:42]: same here 
```suggestion
    private final String hostIP;
```

**Review by DaanHoogland — APPROVED** [review]
code looks good (in spite of the unguarded formats in debug statements. but this needs (as said) extensive 3rd party testing.

**Comment by DaanHoogland** [comment]
> Can anyone review this PR?

I just did, but this is not what you need most, I'll look in the company to find test effort for us, to get this merged. As a general rule we need two lgtm, on of which should have at least tested extensively and the other at least looked at the code.

**Comment by GutoVeronezi** [comment]
@blueorangutan test

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java
Comment (GabrielBrascher) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:42]: If the contributor decides to proceed with renaming this variable, I would also suggest removing the commented line `/* private IP address */`, and leave simply `hostIP`, or then something as `hostPrivateIP` or `privateIpAddress`, as this is expected to be the `HostVO.privateIpAddress`.

**Review by GabrielBrascher — APPROVED** [review]
Code LGTM.

I added a suggestion regarding the variable naming change; nothing that would block this PR anyway.

**Comment by GutoVeronezi** [comment]
@DaanHoogland @GabrielBrascher rename done.

File: agent/conf/agent.properties
Comment (sureshanaparti) [inline agent/conf/agent.properties:268]: typo `hearbeat` => _heartbeat_

File: agent/src/main/java/com/cloud/agent/properties/AgentProperties.java
Comment (sureshanaparti) [inline agent/src/main/java/com/cloud/agent/properties/AgentProperties.java:34]: check the typo error 'hearbeat', and correct it wherever applicable.

**Comment by sureshanaparti** [comment]
@GutoVeronezi is this PR ready for review?

**Comment by GutoVeronezi** [comment]
@sureshanaparti as I mentioned in https://github.com/apache/cloudstack/pull/4585#issuecomment-850866396, this PR will be rebased if https://github.com/apache/cloudstack/pull/4585 be merged, therefore it still is a WiP.

**Comment by GutoVeronezi** [comment]
@DaanHoogland @GabrielBrascher @RodrigoDLopez @sureshanaparti 
As https://github.com/apache/cloudstack/pull/4585 was merged, I rebased this one and applied only the real changes.

**Comment by DaanHoogland** [comment]
@GutoVeronezi we may need to revert #4585 as addHost is failing since it got merged. Not sure what is the issue, still investigating.

**Comment by GutoVeronezi** [comment]
Converting it to draft due to #4585 revertion.

**Comment by GutoVeronezi** [comment]
@DaanHoogland @GabrielBrascher @RodrigoDLopez @sureshanaparti
As #5239 was merged, I rebased this one and applied only the real changes.

**Comment by GutoVeronezi** [comment]
@blueorangutan package

**Review by RodrigoDLopez — COMMENTED** [review]
The code LGTM.
i'll be able to test this one on this weekend. perhaps i'll post my results in the next monday.

File: agent/conf/agent.properties
Comment (DaanHoogland) [inline agent/conf/agent.properties:275]: hope the newline is there ;)

**Review by DaanHoogland — COMMENTED** [review]
cltgm

File: agent/conf/agent.properties
Comment (GutoVeronezi) [reply agent/conf/agent.properties:275]: @DaanHoogland It's there =)

GitHub warns when there is no newline at the end of the file, e.g.:
image

Comment (DaanHoogland) [reply agent/conf/agent.properties:275]: ah, you're right. forgot about that.

**Comment by DaanHoogland** [comment]
test_vpc_redundant is failing all over the open PRs, but running one last time anyway to be as sure as possible

**Comment by DaanHoogland** [comment]
@RodrigoDLopez did you test this?

**Comment by RodrigoDLopez** [comment]
> @RodrigoDLopez did you test this?

Sorry @DaanHoogland I couldn't run the tests yet. I'm having problems with my test environment, but I will run the tests as soon as possible.

**Comment by GutoVeronezi** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
> > @RodrigoDLopez did you test this?
> 
> Sorry @DaanHoogland I couldn't run the tests yet. I'm having problems with my test environment, but I will run the tests as soon as possible.

@RodrigoDLopez anything to share yet?

**Comment by RodrigoDLopez** [comment]
I ran some tests with this propose. But I'm not sure about my environment. I couldn't see my kvm agent stop at any point.
anyway, I managed to verify that the behavior of the externalized parameter is in accordance with what was proposed.

forcing my heartbeat timeout on two kvm hosts it's possible to see that the ones without the parameter `reboot.host.and.alert.management.on.heartbeat.timeout=false` will  log the mensage `failed: timeout; stopping cloudstack-agent.`

**Kvm-nd01**
```
heartbeat.update.timeout=10
reboot.host.and.alert.management.on.heartbeat.timeout=true
```
```
Thread-1:null) (logid:) The command (/usr/share/cloudstack-common/scripts/vm/hypervisor/kvm/kvmheartbeat.sh -i 201.200.102.10 -p /export/primary -m /mnt/0727b905-35cb-36cb-961c-ea27b1a83333 -h 201.200.101.3 ), to the pool [0727b905-35cb-36cb-961c-ea27b1a83333], has the result [timeout].
(Thread-1:null) (logid:) Write heartbeat for pool [0727b905-35cb-36cb-961c-ea27b1a83333] failed: timeout; try: 5 of 5.
(Thread-1:null) (logid:) [IGNORED] Interrupted between heartbeat retries.
java.lang.InterruptedException: sleep interrupted
	at java.base/java.lang.Thread.sleep(Native Method)
	at com.cloud.hypervisor.kvm.resource.KVMHAMonitor.runHeartBeat(KVMHAMonitor.java:129)
	at com.cloud.hypervisor.kvm.resource.KVMHAMonitor.run(KVMHAMonitor.java:176)
	at java.base/java.lang.Thread.run(Thread.java:829)
(Thread-1:null) (logid:) Write heartbeat for pool [0727b905-35cb-36cb-961c-ea27b1a83333] failed: timeout; stopping cloudstack-agent. 
```

**Kvm-nd02**
```
heartbeat.update.timeout=10
reboot.host.and.alert.management.on.heartbeat.timeout=false
```
```
(Thread-1:null) (logid:) The command (/usr/share/cloudstack-common/scripts/vm/hypervisor/kvm/kvmheartbeat.sh -i 201.200.102.10 -p /export/primary -m /mnt/0727b905-35cb-36cb-961c-ea27b1a83333 -h 201.200.101.4 ), to the pool [0727b905-35cb-36cb-961c-ea27b1a83333], has the result [timeout].
(Thread-1:null) (logid:) Write heartbeat for pool [0727b905-35cb-36cb-961c-ea27b1a83333] failed: timeout; try: 5 of 5.
(Thread-1:null) (logid:) [IGNORED] Interrupted between heartbeat retries.
java.lang.InterruptedException: sleep interrupted
	at java.base/java.lang.Thread.sleep(Native Method)
	at com.cloud.hypervisor.kvm.resource.KVMHAMonitor.runHeartBeat(KVMHAMonitor.java:129)
	at com.cloud.hypervisor.kvm.resource.KVMHAMonitor.run(KVMHAMonitor.java:176)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

**Review by GabrielBrascher — APPROVED** [review]
Thanks for the PR @GutoVeronezi.

LGTM based on code review and manual tests done by @RodrigoDLopez.

The test looks good, probably the host was not rebooted due to some test env details. The logic behind loading the agent configuration, validating it and changing the execution flow according to it seems OK.

**Comment by GutoVeronezi** [comment]
@DaanHoogland @GabrielBrascher @RodrigoDLopez is this good to go? Is there anything else to do?

**Comment by nvazquez** [comment]
Thanks for testing @RodrigoDLopez 
@GutoVeronezi I think we can kick a new round of tests then should be good to go
@blueorangutan package

**Comment by GutoVeronezi** [comment]
@nvazquez ok, thanks!

**Comment by nvazquez** [comment]
@blueorangutan test

**Review by nvazquez — APPROVED** [review]

**Pull request merged**
